### ISSUE TITLE
Add preference for long args/`when` in methods?

### ISSUE THREAD

**Issue by dylan-chong** [christopheradams/elixir_style_guide#162]
# Situation 1 (short def, with `when`)

I find that if a method has long args or lots of `when` conditions, the method will have to go across multiple lines. 
I personally prefer doing something like this for example (2 indents for the `when`s):

```
  def meth(a, b, c)
      when cond_1
      and cond_2
      and cond_3
      and cond_4
  do
    some
    stuff
  end
```

# Situation 2 (long def, without `when`)

---

With a long method definition with no when statement. 
Here I prefer having 2 indents for the params.

```
  def very_very_very_very_very_long_method_name(
      lots,
      of,
      parameters,
      a,
      b,
      c,
      d,
      e,
  ) do
    some    
    stuff
  end
```

---

# Situation 3 (long def, with `when`)

A combination of the above 2 situations.

```
  def very_very_very_very_very_long_method_name(
      lots,
      of,
      parameters,
      a,
      b,
      c,
      d,
      e,
  ) when cond_1
      and cond_2
      and cond_3
      and cond_4
  do
    some    
    stuff
  end
```

---

Might something like above situations (whether you follow my preference or not) be added to the style guide?
I also would like to hear about other people's preferences for the above
situation

**Comment by christopheradams** [comment]
You're right that the guide is missing some explicit rules for function heads and guards that don't fit on a single line.

Fortunately the new Elixir code formatter will handle all of these cases. I re-wrote your examples to trigger line breaking and ran `mix format` over them:

```elixir
  def meth(a, b, c)
      when very_long_condition1 and very_long_condition2 and very_long_condition3 and
             very_long_condition4 do
    some
    stuff
  end

  def very_very_very_very_very_long_method_name(
        lots,
        of,
        parameters,
        a,
        b,
        c,
        d,
        e,
        etc,
        etc,
        etc
      ) do
    some
    stuff
  end

  def very_very_very_very_very_long_method_name(
        lots,
        of,
        parameters,
        a,
        b,
        c,
        d,
        e,
        etc,
        etc,
        etc
      )
      when very_long_condition1 and very_long_condition2 and very_long_condition3 and
             very_long_condition4 do
    some
    stuff
  end
```

**Comment by dylan-chong** [comment]
Interesting, I quite like that style except for the conditions being forced to take up the entire 100 characters. I hope theres an option to change that.

Is there some new format are built into mix itself?

**Comment by christopheradams** [comment]
The formatter will be available starting in Elixir 1.6. However, there are no options for configuring the above line-breaking rules.
